### PR TITLE
[cmake] use wfn91's linear algebra discovery modules

### PR DESCRIPTION
**Pull request by evaleev**

### CODE REVIEW

**Comment by wavefunction91** [comment]
@evaleev This passes all CI except for Travis + clang (which is not a problem in the build, it's a problem in the Ubuntu packages). Do we want to fix the Travis build, or take Gitlab as being indicative of it working?

**Comment by evaleev** [comment]
@wavefunction91 I think this is ready. I'm pretty sure my updates of the dox (INSTALL.md) are not 100% correct (does your kit respect `BLA_STATIC`?), but are mostly there

**Comment by wavefunction91** [comment]
We have `BLAS_PREFERS_STATIC`, but I can add a `BLA_STATIC` compatibility if you'd prefer

File: INSTALL.md [hunk 1]
```diff
@@ -230,38 +230,76 @@ set the environment variables (e.g. `OMP_NUM_THREADS`, `MKL_NUM_THREADS`, etc.)
...
+To discover and configure the use of Intel MKL consider these suggestions:
+- The use of NWChemEx discovery kit is strongly recommended for discovering Intel MKL. The following CMake cache variables can be used to specify the desired Intel MKL configuration:
+  - `intelmkl_PREFERS_STATIC`: whether to look for static or shared/dynamic libraries (default = `OFF`)
```
Comment (wavefunction91) [inline INSTALL.md:292]: This would be covered by `BLAS_PREFERS_STATIC`

File: INSTALL.md [hunk 2]
```diff
@@ -230,38 +230,76 @@ set the environment variables (e.g. `OMP_NUM_THREADS`, `MKL_NUM_THREADS`, etc.)
...
+  imported transitively via the BTAS library). Under the most common scenario, where TiledArray
+  will configure and compile BTAS dependency and its BLAS++/LAPACK++ prerequisites from source
+  (this is strongly recommended), BLAS/LAPACK will thus be discovered and imported by
+  BLAS++/LAPACK++ during the TA configuration. There are 2 mechanisms by which BLAS++/LAPACK++
+  discover BLAS/LAPACK:
+  - _the built-in custom discovery kit_; no options exist to provide any control
+  - standard CMake BLAS/LAPACK modules.
+  
+  The latter is used if CMake cache variable `BLA_VENDOR` is specified:
+  - `BLA_VENDOR` -- controls which vendor BLAS/LAPACK library will be sought
```
[the remark this answers is not in the document]
Comment (evaleev) [reply INSTALL.md:269]: I think it is least surprising to emulate `Find{BLAS,LAPACK}` as much as reasonable, no?

File: INSTALL.md [hunk 3]
```diff
@@ -230,38 +230,76 @@ set the environment variables (e.g. `OMP_NUM_THREADS`, `MKL_NUM_THREADS`, etc.)
...
-[LAPACK++](https://icl.bitbucket.io/lapackpp/md__i_n_s_t_a_l_l.html).
+  More information can be found in the installation instructions for
+  [BLAS++](https://icl.bitbucket.io/blaspp/md__i_n_s_t_a_l_l.html) and
+  [LAPACK++](https://icl.bitbucket.io/lapackpp/md__i_n_s_t_a_l_l.html).
+  
+  Note that BLAS++/LAPACK++ discover BLAS and LAPACK only; ScaLAPACK
+  library is always discovered using the NWChemEx kit.
+
+Also note that all discovery methods respect the following CMake cache variable:
+- `BLA_STATIC` -- indicates whether static or shared LAPACK and BLAS libraries will be preferred.
```
[the remark this answers is not in the document]
Comment (evaleev) [reply INSTALL.md:282]: ditto

File: INSTALL.md (hunk 2 above)
[the remark this answers is not in the document]
Comment (evaleev) [reply INSTALL.md:269]: i.e. I think it's reasonable to read same input vars as those, i.e. read `BLA_{VENDOR,STATIC}`

[the remark this answers is not in the document]
Comment (wavefunction91) [reply INSTALL.md:269]: I suppose that is the least surprising behaviour, I'll patch that today

**Comment by wavefunction91** [comment]
@evaleev Kitware compatibility is added in https://github.com/wavefunction91/linalg-cmake-modules/pull/1/commits/510b9282faa795925f942e8d7d0060b2d8ea9668 , we should just need to bump the version in `vg-kit`. This will allow for `BLA_STATIC` -> `{BLAS,LAPACK,ScaLAPACK}_PREFERS_STATIC` emulation. Same goes for `BLA_VENDOR`, but as we've discussed offline, this generally leads to ambiguity in the cases where BLAS/LAPACK interoperability is desired

The following are the allowed specs for `XXX_PREFERENCE_LIST`:

`XXX=BLAS`
- `ReferenceBLAS` (NETLIB)
- `IntelMKL`
- `IBMESSL`
- `BLIS`
- `OpenBLAS`
- `Accelerate`

`XXX=LAPACK`
- `ReferenceLAPACK` (NETLIB)
- `FLAME`

**Comment by evaleev** [comment]
@wavefunction91 OK, thanks, will do

**Comment by evaleev** [comment]
> @evaleev Kitware compatibility is added in [wavefunction91/linalg-cmake-modules@510b928](https://github.com/wavefunction91/linalg-cmake-modules/commit/510b9282faa795925f942e8d7d0060b2d8ea9668) , we should just need to bump the version in `vg-kit`.

[done](https://github.com/ValeevGroup/kit-cmake/commit/a09f7905e4359cacea307d7e2fa945fb5c03400a)

> 
> The following are the allowed specs for `XXX_PREFERENCE_LIST`:
> 
> `XXX=BLAS`
> 
> * `ReferenceBLAS` (NETLIB)
> * `IntelMKL`
> * `IBMESSL`
> * `BLIS`
> * `OpenBLAS`
> * `Accelerate`
> 
> `XXX=LAPACK`
> 
> * `ReferenceLAPACK` (NETLIB)
> * `FLAME`

Does the user need to specify `{LAPACK,ScaLAPACK}_PREFERENCE_LIST` in the case `BLAS_PREFERENCE_LIST= IntelMKL`? I.e. does the latter imply that only `IntelMKL` will be considered when searching for LAPACK and ScaLAPACK?

**Comment by wavefunction91** [comment]
@evaleev The user should not have to specify `{Sca,}LAPACK_PREFERENCE_LIST` for MKL. Take ScaLAPACK for example (since its the highest up on the stack) - if unguided discovery is used, then it will go like

1. FindScaLAPACK -> no `ScaLAPACK_LIBRARIES` are set, so it looks for LAPACK
2. FindLAPACK -> no `LAPACK_LIBRARIES` are set, looks for BLAS
3. FindBLAS -> no `BLAS_LIBRARIES` are set, goes through `BLAS_PREFERENCE_LIST`, finds `IntelMKL`, propagates `IntelMKL_LIBRARIES` to `BLAS_LIBRARIES`. Because ScaLAPACK was requested, it also tacks on the BLACS/ScaLAPACK MKL linker (+ MPI), which is non-trivial (which is why it's important to ask for the highest level of the stack desired so it can propagate down to `Find<VENDOR>` if need be)
4. FindLAPACK -> recognizes that it found BLAS, checks for LAPACK linker in `BLAS_LIBRARIES`, this will be successful for MKL and will propagate `BLAS_LIBRARIES` (and associated variables) to `LAPACK_LIBRARIES`
5. FindScaLAPACK -> recognizes that it found LAPACK, checks for ScaLAPACK linker in `LAPACK_LIBRARIES` (which will work for MKL) and propagates to `ScaLAPACK_LIBRARIES`, etc.

**Comment by evaleev** [comment]
@wavefunction91 sorry for wasting your time ... since the proper thing will be to interpret `BLA_VENDOR` rather than just passing it on as `BLAS_PREFERENCE_LIST` it's probably best to just use your kit's vars directly ... only realized this after seeing what happens when `BLA_VENDOR` used with standard CMake module is used with your kit: https://gitlab.com/ValeevGroup/tiledarray/-/jobs/1291849639#L538  ... I propose we avoid the use of BLA_VENDOR to avoid these kinds of issues.

**Pull request merged**